plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/Utils.hs

```diff
@@ -16,86 +16,153 @@ import PlutusIR.Purity (firstEffectfulTerm, isPure)
 import Control.Monad.Reader
 import Control.Monad.State
 
 import Data.Map qualified as Map
 import Data.Semigroup.Generic (GenericSemigroupMonoid (..))
+import Prettyprinter (viaShow)
+
+-- For unconditional inlining:
 
 -- | Substitution range, 'SubstRng' in the paper but no 'Susp' case.
 -- See Note [Inlining approach and 'Secrets of the GHC Inliner']
-newtype InlineTerm tyname name uni fun a =
-    Done (Dupable (Term tyname name uni fun a)) --out expressions
+newtype InlineTerm tyname name uni fun ann =
+    Done (Dupable (Term tyname name uni fun ann)) --out expressions
 
 -- | Term substitution, 'Subst' in the paper.
 -- A map of unprocessed variable and its substitution range.
-newtype TermEnv tyname name uni fun a =
-    TermEnv { _unTermEnv :: UniqueMap TermUnique (InlineTerm tyname name uni fun a) }
+newtype TermEnv tyname name uni fun ann =
+    TermEnv { _unTermEnv :: UniqueMap TermUnique (InlineTerm tyname name uni fun ann) }
     deriving newtype (Semigroup, Monoid)
 
 -- | Type substitution, similar to `TermEnv` but for types.
 -- A map of unprocessed type variable and its substitution range.
-newtype TypeEnv tyname uni a =
-    TypeEnv { _unTypeEnv :: UniqueMap TypeUnique (Dupable (Type tyname uni a)) }
+newtype TypeEnv tyname uni ann =
+    TypeEnv { _unTypeEnv :: UniqueMap TypeUnique (Dupable (Type tyname uni ann)) }
     deriving newtype (Semigroup, Monoid)
 
+-- For call site inlining:
+
+-- | A mapping including all let-bindings that are functions.
+newtype CalledVarEnv tyname name uni fun ann =
+    CalledVarEnv { _unCalledVarEnv :: UniqueMap TermUnique (CalledVarInfo tyname name uni fun ann)}
+    deriving newtype (Semigroup, Monoid)
+
+{-|
+The (syntactic) arity of a term. That is, a record of the arguments that the
+term expects before it may do some work. Since we have both type and lambda
+abstractions, this is not a simple argument count, but rather a list of values
+indicating whether the next argument should be a term or a type.
+
+Note that this is the syntactic arity, i.e. it just corresponds to the number of
+syntactic lambda and type abstractions on the outside of the term. It is thus
+an under-approximation of how many arguments the term may need.
+e.g. consider the term @let id = \x -> x in id@: the variable @id@ has syntactic
+arity @[]@, but does in fact need an argument before it does any work.
+-}
+type Arity = [TermOrType]
+
+-- | An ordered list of the term and type arguments applied to a function.
+type AppOrder = [TermOrType]
+
+-- | Info attached to a let-binding needed for call site inlining.
+data CalledVarInfo tyname name uni fun ann =
+  MkCalledVarInfo {
+    calledVarDef    :: Term tyname name uni fun ann -- ^ its definition
+    , arity         :: Arity -- ^ its sequence of term and type lambdas
+    , calledVarBody :: Term tyname name uni fun ann
+    -- ^ the body of the function, for checking `acceptable` or not
+  }
+-- | Is the next argument a term or a type?
+data TermOrType =
```

Comment:
Naming suggestion. At the moment you have two lists of things that get kind of described as "arg lists": the arity, and the arguments at the call site. I think the usual CS terminology is to call the first thing "parameters" and the second thing "arguments". So how about:
```
data ParamKind
  TermParam | TypeParam
```